src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentStateCopy = { ...state };
+
+  for (const action of actions) {
+    const stateCopy = { ...currentStateCopy };
+
+    if (action.type === 'addProperties') {
+      for (const [key, value] of Object.entries(action.extraData)) {
+        stateCopy[key] = value;
+      }
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete stateCopy[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      for (const key in stateCopy) {
+        delete stateCopy[key];
+      }
+    }
+
+    result.push(stateCopy);
```

Comment:
[CODE STYLE]: You should push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after the current action is processed. This ensures that the same copy isn't modified by multiple actions.